design/external-authorization-design.md

```diff
@@ -0,0 +1,670 @@
+	// +required
+	Kind string `json:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`
+	// Namespace of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,opt,name=namespace"`
+	// Name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	// +required
+	Name string `json:"name,omitempty" protobuf:"bytes,3,opt,name=name"`
+}
+
+type AuthorizationPolicy struct {
+	// When true, this field disables client request authentication
+	// for the scope of the policy.
+	//
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
+
+	// Context is a set of key/value pairs that are sent to the
+	// authentication server in the check request. If a context
+	// is provided at an enclosing scope, the entries are merged
+	// such that the inner scope overrides matching keys from the
+	// outer scope.
+	//
+	// +optional
+	Context map[string]string `json:"context,omitempty"`
+}
+
+type AuthorizationServer struct {
+	// +required
+	ServiceRef SupportServiceReference `json:"serviceRef"`
+
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
+
+	// Timeout configures maximum time to wait for a check response from the authorization server.
+	// Timeout durations are expressed in the Go [Duration format](https://godoc.org/time#ParseDuration).
+	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+	// The string "infinity" is also a valid input and specifies no timeout.
+	//
+	// +optional
+	Timeout string `json:"timeout,omitempty"`
+
+    // If FailOpen is true, the client request is forwarded to the upstream service
+    // even if the authorization server fails to respond. This field should not be
+    // set in most cases. It is intended for use only while migrating applications
+    // from internal authorization to Contour external authorization.
+    //
+    // +optional
+    FailOpen bool `json:"failOpen,omitempty"`
+}
+
+// VirtualHost appears at most once. If it is present, the object is considered
+// to be a "root".
+type VirtualHost struct {
+    ...
+
+	// This field configures a support service to perform
+	// authorization for this virtual host. Authorization can
+	// only be configured on virtual hosts that have TLS enabled.
+	// If the TLS configuration requires client certificate
+	///validation, the client certificate is always included in the
+	// authentication check request.
+	//
+	// +optional
+	Authorization *AuthorizationServer `json:"authorization,omitempty"`
+}
+
+// Route contains the set of routes for a virtual host.
+type Route struct {
+    ...
+
+    // AuthPolicy updates the authorization policy for client
+	// requests that match this route.
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
```

Comment:
Can I have an AuthorizationPolicy in an include? I'm thinking of setting a bunch of routes that should be insecure, you could include them and add this to the includes conditions without having to set them on each child route.

Reply:
Does anything else in the HTTPProxy API behaves that way? If there's an established pattern we can try to follow it here, but I don't think that this should be too different from what we already have.

Reply:
We could add a `AuthorizationPolicy` to the include and I can see the usefulness of that. But it's more complexity, so I'd rather wait and address it when there's demand. I'm expecting that `AuthorizationPolicy` won't be tweaked that much in practice.

Reply:
Just something to think about. By doing this, you'd have an easy way to not need the route disable bit. You could put the insecure routes in an HTTPProxy chain, then the secure in another.

Reply:
Yeh, just a matter of whether we want to do this up front in what will already be a big change.

Reply:
I agree that I think we leave this one and see what our users want. We're not ruling this out by not doing it now.